feat: implementation

MQ37 opened the pull request:
Supports normal and standby mode.

Image size: 163.4 MB.

Supported LLM models:
```
                "google/gemini-2.5-flash",
                "google/gemini-2.0-flash-001",
                "openai/gpt-4.1",
                "openai/gpt-4.1-mini"
```

related to https://github.com/apify/ai-team/issues/14

MQ37 committed:
initial snapshot

MQ37 committed:
add actor context

MQ37 committed:
Refactor LLM query functions and add query sanity check

MQ37 committed:
Update queryLLM to return detailed response with token usage

MQ37 committed:
clean code, linter rules, improve querying - add important fields reasoning LLM query

MQ37 committed:
use apify open router proxy, switch to gpt 4.1 models

MQ37 committed:
simple standby mode

MQ37 committed:
use gemini flash models, they are faster

MQ37 committed:
lint, format

MQ37 committed:
update readme

MQ37 committed:
improve input

jirispilka commented:
Do you think it would be worth exposing this as an MCP server as well?
It could help promote MCP servers and also contribute to our KR 😊

For example, we could wrap the standby server using [fastmcp](https://github.com/punkpeye/fastmcp)

MQ37 commented:
> Do you think it would be worth exposing this as an MCP server as well? It could help promote MCP servers and also contribute to our KR 😊
> 
> For example, we could wrap the standby server using [fastmcp](https://github.com/punkpeye/fastmcp)

Great idea :+1: Let's go all in - normal, standby and MCP :+1:

MQ37 committed:
mcp server and endpoints, oauth support, remove the query sanity check step

MQ37 commented:
@jirispilka Added MCP support (also with OAauth, we should add oauth also to MCP templates - issue https://github.com/apify/actor-templates/issues/486). Also removed the query sanity check, because in some cases it was not working properly and removing it makes the Actor faster (LLM calls take quite a bit of time)

jirispilka commented on .actor/actor.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "actorSpecification": 1,
+    "name": "tiny-dataset-query-engine",
```
Let's add description tag as well

jirispilka commented on .actor/actor.json:
```diff
@@ -0,0 +1,12 @@
+        "dataset": "./dataset_schema.json"
+    },
+    "input": "./input_schema.json",
+    "dockerfile": "./Dockerfile"
```
"webServerMcpPath": "/mcp" is missing

jirispilka commented on src/standby.ts:
```diff
@@ -0,0 +1,243 @@
+        }
+    });
+
+    //=============================================================================
```
I would remove legacy SSE. Let's go with http-streamable only. Since it is new we don't have to be backwards compatible

MQ37 replied:
I understand but someone might want to connect directly to this Actor with older MCP client and it might not work with the streamable. Since it is already there and works I would keep it.

jirispilka commented on .actor/input_schema.json:
```diff
@@ -0,0 +1,41 @@
+        "modelName": {
+            "title": "OpenAI model",
+            "type": "string",
+            "description": "The OpenAI model to use. Currently supported models are Gemini 2.5 Flash, Gemini 2.0 Flash, GPT-4.1, and GPT-4.1 Mini.",
```
We are using open-router, so we should be ablet use more models. Is there a reason why to restrict it to the selected four?

MQ37 replied:
I restricted to a few models that I tested, I think the end goal here is to remove this field and make this a black box that just works without the user overhead of choosing a model. Or we can change the input to something like `smart` and `fast` mode (to choose between smarted, slower and more expensive vs faster, cheaper and not so smart model)

jirispilka commented on src/engine.ts:
```diff
@@ -0,0 +1,176 @@
+    const dataset = (await getApifyDataset(input.dataset)) as { actId: string };
+    const datasetItems = await getApifyDatasetItems(input.dataset);
+
+    log.info("infering dataset shape...");
```
```suggestion
    log.info("inferring dataset shape...");
```

jirispilka commented on src/engine.ts:
```diff
@@ -0,0 +1,176 @@
+export async function runQuery(input: Input): Promise<string> {
+    const model = input.modelName;
+
+    log.info("getting dataset...");
```
```suggestion
    log.info("Getting dataset...");
```

jirispilka commented on src/engine.ts:
```diff
@@ -0,0 +1,176 @@
+    const tableShape = convertTypeShapeToTableShape(typeShape);
+    log.debug(`Table shape: ${JSON.stringify(tableShape)}`);
+
+    log.info("preparing database engine...");
```
```suggestion
    log.info("Preparing database engine...");
```

jirispilka commented on src/index.ts:
```diff
@@ -0,0 +1,42 @@
+    const input = await handleInput();
+    if (!input) {
+        log.error("Input is invalid");
+        await Actor.exit({ statusMessage: "Input is invalid" });
```
It would be nice to also return the error.
Or at least tell the user to check logs

MQ37 replied:
Lets throw, catch it, log it and tell user to check the logs :+1:

jirispilka commented on src/dataset.ts:
```diff
@@ -0,0 +1,103 @@
+    const datasetClient = apifyClient.dataset(id);
+
+    const items = (await datasetClient.listItems()).items;
+    return items;
```
### 
```suggestion
    return (await datasetClient.listItems()).items;
```

jirispilka commented on src/dataset.ts:
```diff
@@ -0,0 +1,103 @@
+    return "unknown";
+}
+
+export function getDatasetTypeShape(
```
can you please document this function

jirispilka commented on src/engine.ts:
```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+export function convertTypeShapeToTableShape(typeShape: TypeShape): TableShape {
```
can you please document it

jirispilka commented on src/llm.ts:
```diff
@@ -0,0 +1,173 @@
+ * @param {TableShape} tableShape - The schema of the table to be analyzed.
+ * @param {ActorContext} actorContext - The context of the actor including name and description.
+ * @returns {Promise<Record<string, string>>} - A promise that resolves to a record where each key is a field name and the value is a short reason for its importance.
+ */
```
```suggestion
/**
 * Queries the LLM to identify important fields from a table schema relevant to a user query.
 *
 * @param {Object} args - The arguments object
 * @param {string} args.prompt - The user query for which important fields need to be identified
 * @param {TableShape} args.tableShape - The schema of the table to be analyzed
 * @param {ActorContext} args.actorContext - The context of the actor including name and description
 * @param {string} [args.model] - The model to use for the LLM query (optional)
 * @returns {Promise<[string, string][]>} - A promise that resolves to an array of tuples where each tuple contains a field name and a short reason for its importance
 */
 */
```

jirispilka commented on README.md:
````diff
@@ -1 +1,36 @@
+
+```text
+https://jakub-kopecky--tiny-dataset-query-engine.apify.actor/?token=YOUR_APIFY_TOKEN&dataset=DATASET_ID&query=YOUR_QUERY
+```
````
You need to add MCP Server usage here

jirispilka commented on src/llm.ts:
```diff
@@ -0,0 +1,173 @@
+import { log } from "apify";
+
+const client = new OpenAI({
+    baseURL: "https:/openrouter.apify.actor/api/v1",
```
should this be in the const?

MQ37 replied:
Agree :+1:

jirispilka commented on src/db.ts:
```diff
@@ -0,0 +1,5 @@
+import { Database } from "bun:sqlite";
+
+export function createDatabase() {
```
consider moving this to engine.ts. I understand this structure ... but I think it is a bit overkill to have a new file for 3 lines of code

MQ37 replied:
Makes sense if it's only just single function and SQLite schema logic is also in the engine

jirispilka commented on README.md:
```diff
@@ -1 +1,36 @@
+
+This Actor uses LLMs to understand your query, generate an SQL statement, and report the results. First, the Actor retrieves the dataset items and builds an internal schema representation, which is converted into an in-memory [SQLite3](https://www.sqlite.org/) database. Then, the LLM generates an SQL query and executes it against the SQLite3 database. Finally, it summarizes the results and return ...
+
+## 🚀 Usage
```
MCP usage is missing

jirispilka replied:
Can you please add more examples? Like user query, generated sql, and result

jirispilka commented on README.md:
```diff
@@ -1 +1,36 @@
+This Actor allows you to run natural language queries against an Apify dataset. For example, if you run a web scraping Actor and want to know how many pages contain a specific keyword, you can ask this Actor to find out for you.
+
+> **ℹ️ Notice:**  
+> This Actor internally uses the [Apify Openrouter Actor](https://apify.com/apify/openrouter) to call an LLM. You will be billed for LLM usage through this Actor.
```
I love that you are using OpenRouter! That was the purpose :)

jirispilka commented on README.md:
```diff
@@ -1 +1,36 @@
+To run the Actor in Standby mode, send an HTTP GET request to the Actor's URL with the following query parameters:
+- `dataset`: The ID of the dataset you want to query.
+- `query`: The natural language query you want to run against the dataset.
+- `modelName`: (optional) The name of the LLM model you want to use for the query. If not provided, the Actor will use the default model defined in the Actor input. See the Actor input schema for a list of available models.
```
we should recommend something, like saying: openai-gpt4-mimi works well for this use-case ....

MQ37 replied:
The Google Gemini models work the best, let's recommend them

jirispilka commented on README.md:
```diff
@@ -1 +1,36 @@
\ No newline at end of file
+# Tiny Dataset Query Engine
+
+This Actor allows you to run natural language queries against an Apify dataset. For example, if you run a web scraping Actor and want to know how many pages contain a specific keyword, you can ask this Actor to find out for you.
```
I think we need to improve this description so that Actor is discoverable. I'm thinking from an MCP perspective

jirispilka reviewed (changes requested):
Looks good. I added a couple of minor comments

MQ37 committed:
address review comments

MQ37 commented:
> Looks good. I added a couple of minor comments

Thank you for review :+1: Addressed the comments. Could you please take a quick look (mainly `README.md`).

jirispilka commented on README.md:
```diff
@@ -9,28 +9,60 @@ This Actor allows you to run natural language queries against an Apify dataset.
+
+> **ℹ️ Notice:**
+> The MCP server supports OAuth authentication. If your client supports it, you can simply connect and go through the OAuth flow to authenticate. If you prefer, or if your client does not support OAuth, you need to either pass the `?token=YOUR_APIFY_TOKEN` query parameter in the request URL or set the `Authorization` header to `Bearer YOUR_APIFY_TOKEN`.
+
```
Looks like a copy paste mistake? duplicated content

MQ37 replied:
Thank you for noticing, I forgot to remove the old part :+1:

jirispilka reviewed (approved):
Looks good! I left a minor comment only.

MQ37 committed:
fix readme
